### PR TITLE
Convert FileStore experiment_id to randomly generated fixed-length int64

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Changes the generation of an experiment_id within FileStore to be a fixed-length random int64 instead of a non-thread-safe monotonically increasing id value. 

## How is this patch tested?

- [X] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

In FileStore, experiment_id generation is now a fixed-length random int64 value. The default experiment_id is still '0' in order to limit the scope of changes.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [X] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/tracking/file_store.py [hunk 1]
```diff
@@ -348,17 +355,14 @@ def create_experiment(self, name, artifact_location=None, tags=None):
...
-            int(e)
-            for e in (
-                self._get_active_experiments(full_path=False)
-                + self._get_deleted_experiments(full_path=False)
+        experiment_id = _generate_unique_integer_id()
+        if not self._has_experiment(FileStore.DEFAULT_EXPERIMENT_ID):
+            warnings.warn(
+                "The default experiment is not present due to a file system issue or manual "
+                "deletion. Recreating the default experiment with "
+                f"`experiment_id={FileStore.DEFAULT_EXPERIMENT_ID}`."
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:363]: Why do we need to create the default experiment in `create_experiment`? Is there an issue if we create a new experiment without the default experiment?

File: mlflow/store/tracking/file_store.py [hunk 2]
```diff
@@ -348,17 +355,14 @@ def create_experiment(self, name, artifact_location=None, tags=None):
...
-        # len(list_all(..)) would not work when experiments are deleted.
-        experiments_ids = [
-            int(e)
-            for e in (
-                self._get_active_experiments(full_path=False)
-                + self._get_deleted_experiments(full_path=False)
+        experiment_id = _generate_unique_integer_id()
+        if not self._has_experiment(FileStore.DEFAULT_EXPERIMENT_ID):
+            warnings.warn(
+                "The default experiment is not present due to a file system issue or manual "
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:361]: If we disallow deleting the default experiment, would that simplify the implementation?

File: mlflow/store/tracking/file_store.py (hunk 1 above)
Comment (BenWilson2) [reply mlflow/store/tracking/file_store.py]: I removed this warning and the call to recreate.

File: mlflow/store/tracking/file_store.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/store/tracking/file_store.py]: We can disallow in the delete experiment API :)

File: tests/utils/test_name_utils.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+def test_random_name_generation():
+
+    # Validate exhausted loop truncation
+    name = _generate_random_name(max_length=8)
+    assert len(name) == 8
+
+    # Validate default behavior while calling 1000 times that names end in integer
+    names = [_generate_random_name() for _ in range(1000)]
+    assert all(len(name) <= 20 for name in names)
```
Comment (harupy) [inline tests/utils/test_name_utils.py:12]:
```suggestion
    assert all(len(name) == 18 for name in names)
```

Can we make this more strict?

File: tests/tracking/test_rest_tracking.py [hunk 4]
```diff
@@ -761,13 +761,14 @@ def test_search_experiments(mlflow_client):
...
     # page_token
     experiments = mlflow_client.search_experiments(page_token=experiments.token)
-    assert [e.name for e in experiments] == ["a", "Default"]
+    assert {e.name for e in experiments}.issubset({"a", "Default"})
 
     # view_type
     mlflow_client.delete_experiment(experiment_ids[1])
     experiments = mlflow_client.search_experiments(view_type=ViewType.ACTIVE_ONLY)
-    assert [e.name for e in experiments] == ["Abc", "a", "Default"]
+    assert {e.name for e in experiments}.issubset({"Abc", "a", "Default"})
```
Comment (harupy) [inline tests/tracking/test_rest_tracking.py:769]: Why do we need to use `issubset` here?

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:769]:
```suggestion
    assert {e.name for e in experiments} == ({"Abc", "a", "Default"})
```

Can we use `==`?

Comment (BenWilson2) [reply tests/tracking/test_rest_tracking.py:769]: good point :) I'll adjust them

File: mlflow/store/tracking/file_store.py [hunk 5]
```diff
@@ -348,17 +354,11 @@ def create_experiment(self, name, artifact_location=None, tags=None):
...
-                + self._get_deleted_experiments(full_path=False)
-            )
-            if e.isdigit()
-        ]
-        experiment_id = max(experiments_ids) + 1 if experiments_ids else 0
+        experiment_id = (
+            _generate_unique_integer_id()
+            if self._has_experiment(FileStore.DEFAULT_EXPERIMENT_ID)
+            else FileStore.DEFAULT_EXPERIMENT_ID
+        )
```
Comment (harupy) [inline mlflow/store/tracking/file_store.py:361]: Can we just do:

```suggestion
        experiment_id = _generate_unique_integer_id()
```

?

File: tests/store/tracking/test_file_store.py [hunk 6]
```diff
@@ -49,6 +50,8 @@ def create_test_run(self):
     def create_experiments(self, experiment_names):
         ids = []
         for name in experiment_names:
+            # ensure that the field `last_update_time` is distinct for search ordering
+            time.sleep(0.05)
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:54]:
```suggestion
            time.sleep(0.001)
```

Can we shorten the sleep length?

File: mlflow/store/tracking/file_store.py (hunk 5 above)
Comment (BenWilson2) [reply mlflow/store/tracking/file_store.py:361]: If we don't generate the default experiment when `start_run()` and it's contained `create_run()` is called within the deferred commit within autologging, the generated experiment_id from `_generate_unique_integer_id()` will not be present in the tracking store, throwing an exception. 
The original implementation also does this https://github.com/mlflow/mlflow/blob/0f062a662e846f28d1fec71137b906e905f9c4a8/mlflow/store/tracking/file_store.py#L390 with the `else 0` statement.

Comment (BenWilson2) [reply mlflow/store/tracking/file_store.py:361]: (see previous CI runs in this PR where I wasn't doing this creation of the default experiment to see the test failures :) )

File: tests/store/tracking/test_file_store.py (hunk 6 above)
Comment (BenWilson2) [reply tests/store/tracking/test_file_store.py:54]: yep!

File: mlflow/store/tracking/file_store.py (hunk 5 above)
Comment (harupy) [reply mlflow/store/tracking/file_store.py:361]: Which test failures? Can you attach attach to thier links?

Comment (harupy) [reply mlflow/store/tracking/file_store.py:361]: This failure?
https://github.com/mlflow/mlflow/actions/runs/3261759552/jobs/5357438063#step:9:6766

Comment (BenWilson2) [reply mlflow/store/tracking/file_store.py:361]: yep!

Comment (harupy) [reply mlflow/store/tracking/file_store.py:361]: I ran `test_autologging_disable_restores_behavior` after applying this suggestion and it passed. Am I missing something?

```diff
diff --git a/mlflow/store/tracking/file_store.py b/mlflow/store/tracking/file_store.py
index 0d3c944cd..877b943cd 100644
--- a/mlflow/store/tracking/file_store.py
+++ b/mlflow/store/tracking/file_store.py
@@ -354,11 +354,7 @@ class FileStore(AbstractStore):
         self._check_root_dir()
         _validate_experiment_name(name)
         self._validate_experiment_does_not_exist(name)
-        experiment_id = (
-            _generate_unique_integer_id()
-            if self._has_experiment(FileStore.DEFAULT_EXPERIMENT_ID)
-            else FileStore.DEFAULT_EXPERIMENT_ID
-        )
+        experiment_id = _generate_unique_integer_id()
         return self._create_experiment_with_id(name, str(experiment_id), artifact_location, tags)
 
     def _has_experiment(self, experiment_id):
```

```
> pytest tests/autologging/test_autologging_utils.py -k test_autologging_disable_restores_behavior
============================================================================================================================================== test session starts ==============================================================================================================================================
platform linux -- Python 3.7.10, pytest-7.1.3, pluggy-1.0.0 -- /home/haru/miniconda3/envs/mlflow-dev-env/bin/python
cachedir: .pytest_cache
rootdir: /home/haru/Desktop/repositories/mlflow, configfile: pytest.ini
plugins: cov-3.0.0, localserver-0.6.0
collected 68 items / 67 deselected / 1 selected                                                                                                                                                                                                                                                                 

tests/autologging/test_autologging_utils.py::test_autologging_disable_restores_behavior PASSED    
```

Comment (harupy) [reply mlflow/store/tracking/file_store.py:361]: @dbczumar @BenWilson2

Why `experiment_id = _generate_unique_integer_id()` didn't work: https://github.com/mlflow/mlflow/pull/7097#discussion_r998886290

Comment (BenWilson2) [reply mlflow/store/tracking/file_store.py]: Thank you for the investigation and fix @harupy :)

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @BenWilson2 ! Tested this out, and it works like a charm!

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**